examples/transactions.c

```diff
@@ -184,135 +154,137 @@ create_transactional_producer (const rd_kafka_topic_partition_t *rktpar) {
         rd_kafka_error_t *error;
         rd_kafka_resp_err_t err;
         rd_kafka_consumer_group_metadata_t *cgmd;
-        rd_kafka_topic_partition_list_t *offset;
+        rd_kafka_topic_partition_list_t *offsets;
 
-        fprintf(stdout, "Committing transaction for %s [%d]\n",
-                state->rktpar->topic, state->rktpar->partition);
+        fprintf(stdout, "Committing transaction\n");
 
         /* Send the input consumer's offset to transaction
          * to commit those offsets along with the transaction itself,
          * this is what guarantees exactly-once-semantics (EOS), that
          * input (offsets) and output (messages) are committed atomically. */
 
-        /* Get the consumer's current group state */
+        /* Get the consumer's current group metadata state */
         cgmd = rd_kafka_consumer_group_metadata(consumer);
 
+        /* Get consumer's current assignment */
+        err = rd_kafka_assignment(consumer, &offsets);
+        if (err || offsets->cnt == 0) {
+                /* No partition offsets to commit because consumer
+                 * (most likely) lost the assignment, abort transaction. */
+                if (err)
+                        fprintf(stderr,
+                                "Failed to get consumer assignment to commit: "
+                                "%s\n", rd_kafka_err2str(err));
+                else
+                        rd_kafka_topic_partition_list_destroy(offsets);
+
+                error = rd_kafka_abort_transaction(consumer, -1);
+                if (error)
+                        fatal_error("Failed to abort transaction", error);
+
+                return 0;
+        }
+
         /* Get consumer's current position for this partition */
-        offset = rd_kafka_topic_partition_list_new(1);
-        rd_kafka_topic_partition_list_add(offset,
-                                          state->rktpar->topic,
-                                          state->rktpar->partition);
-        err = rd_kafka_position(consumer, offset);
+        err = rd_kafka_position(consumer, offsets);
         if (err)
-                fatal("Failed to get consumer position for %s [%d]: %s",
-                      state->rktpar->topic, state->rktpar->partition,
+                fatal("Failed to get consumer position: %s",
                       rd_kafka_err2str(err));
 
         /* Send offsets to transaction coordinator */
-        error = rd_kafka_send_offsets_to_transaction(state->producer,
-                                                     offset, cgmd, -1);
+        error = rd_kafka_send_offsets_to_transaction(producer,
+                                                     offsets, cgmd, -1);
         rd_kafka_consumer_group_metadata_destroy(cgmd);
-        rd_kafka_topic_partition_list_destroy(offset);
+        rd_kafka_topic_partition_list_destroy(offsets);
         if (error) {
                 if (rd_kafka_error_txn_requires_abort(error)) {
```

Comment:
it's mandatory to check an error returned by a transactional api to see whether it requires and abort (right?), and then abort if so. why don't the methods just do this, and return an error indicating that they've done it? would have lead to a simpler to user api. or am i missing a benefit?

Reply:
mimicing the Java API, afair